### PR TITLE
Cancel Kubernetes Job when CancelScript is called

### PR DESCRIPTION
**Pull request by APErebus**
# Background

Currently, when a script being executed via Kubernetes Job is cancelled (via the CancelScript RPC call), we don't cancel the Job.

# Results

In the `RunningKubernetesJob`, we now attach a cancellation callback to the `scriptCancellationToken`, so we are notified when the cancellation token is cancelled.

This callback spawns a background task which calls the `DeleteJob` method in the `KubernetesJobService`. This marks the job for termination, which triggers a `SIGTERM` signal, allowing for graceful shutdown before the termination command is sent 30s later (`SIGKILL`).

I have adjusted the code so that the `scriptCancellationToken` only cancels the job, not the monitoring of the job (which will now stop watching when the job is deleted)

image

Shortcut story: [sc-67104]

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Octopus.Tentacle/Kubernetes/Scripts/RunningKubernetesJob.cs [hunk 1]
```diff
@@ -165,7 +179,13 @@ async Task<int> CheckIfJobHasCompleted(CancellationToken cancellationToken, Canc
             {
                 log.Error(ex);
                 resultStatusCode = 0;
-            }, cancellationToken);
+            }, CancellationToken.None);
+
+            //if the job was killed by cancellation, then we need to change the exit code
```
Comment (MissedTheMark) [inline source/Octopus.Tentacle/Kubernetes/Scripts/RunningKubernetesJob.cs:184]: Good pick-up, would have been easy to forget this 🥇

**Review by MissedTheMark — APPROVED** [review]
Looks good!

File: source/Octopus.Tentacle/Kubernetes/Scripts/RunningKubernetesJob.cs [hunk 2]
```diff
@@ -59,41 +63,68 @@ public class RunningKubernetesJob : IRunningScript
...
+                    writer.WriteOutput(ProcessOutputSource.StdOut, "Script execution canceled.");
+                    writer.WriteVerbose($"Cancelling Kubernetes Job '{jobName}'.");
+
+                    //we spawn the pod cancellation on a background thread (as this callback runs synchronously)
+                    Task.Run(async () =>
+                    {
+                        try
+                        {
+                            //to cancel the job we just kill the underlying pod(s)
+                            await podService.DeletePodsForJob(jobName, CancellationToken.None);
```
Comment (APErebus) [inline source/Octopus.Tentacle/Kubernetes/Scripts/RunningKubernetesJob.cs:99]: To cancel the job, we need to delete the pod

File: source/Octopus.Tentacle/Kubernetes/Scripts/KubernetesJobScriptExecutor.cs [hunk 3]
```diff
@@ -29,10 +29,7 @@ public IRunningScript ExecuteOnBackgroundThread(StartScriptCommandV3Alpha comman
         {
             var runningScript = new RunningKubernetesJob(workspace, workspace.CreateLog(), command.ScriptTicket, command.TaskId, cancellationToken, log, scriptStateStore, jobService, containerResolver, appInstanceSelector);
```
[the remark this answers is not in the document]
Comment (APErebus) [reply source/Octopus.Tentacle/Kubernetes/Scripts/KubernetesJobScriptExecutor.cs]: Yeah, good call

[the remark this answers is not in the document]
Comment (APErebus) [reply source/Octopus.Tentacle/Kubernetes/Scripts/KubernetesJobScriptExecutor.cs]: fixed in [e1fbeed](https://github.com/OctopusDeploy/OctopusTentacle/pull/742/commits/e1fbeed6016e5a24ea6ee20d1d1cf1edc5ad9041)

File: source/Octopus.Tentacle/Kubernetes/Scripts/RunningKubernetesJob.cs (hunk 2 above)
Comment (APErebus) [reply source/Octopus.Tentacle/Kubernetes/Scripts/RunningKubernetesJob.cs]: Changing to suspending the job, then deleting it

**Comment by APErebus** [comment]
After advice/discussion with @cailyoung, I have changed it again so that the job is suspended (which tells the control plane to terminate the pods), then we delete the job

File: source/Octopus.Tentacle/Kubernetes/Scripts/RunningKubernetesJob.cs [hunk 4]
```diff
@@ -65,35 +66,60 @@ public class RunningKubernetesJob : IRunningScript
...
+                    //we spawn the job cancellation on a background thread (as this callback runs synchronously)
+                    Task.Run(async () =>
+                    {
+                        try
+                        {
+                            writer.WriteVerbose($"Cancelling Kubernetes Job '{jobName}'.");
+                            //first we suspend the job, which terminates the underlying pods
+                            await jobService.SuspendJob(scriptTicket, CancellationToken.None);
+                            //then we delete the job (because we no longer need it)
+                            await jobService.Delete(scriptTicket, CancellationToken.None);
```
Comment (MissedTheMark) [inline source/Octopus.Tentacle/Kubernetes/Scripts/RunningKubernetesJob.cs:95]: +1 for helpful comments like this 👍

Comment (APErebus) [inline source/Octopus.Tentacle/Kubernetes/Scripts/RunningKubernetesJob.cs:95]: new logic is that we suspend the job, then delete it
> When you suspend a Job, any running Pods that don't have a status of Completed will be [terminated](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination). with a SIGTERM signal.

**Review by MissedTheMark — APPROVED** [review]
New change looks to be a lot more graceful 👍

**Pull request merged**